server/test/controllers/api/ProgramApplicationsApiControllerTest.java

```diff
@@ -0,0 +1,142 @@
+import com.jayway.jsonpath.DocumentContext;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Optional;
+import models.ApiKey;
+import models.Program;
+import org.apache.http.HttpStatus;
+import org.junit.Before;
+import org.junit.Test;
+import play.mvc.Http;
+import play.mvc.Result;
+import services.applicant.JsonPathProvider;
+import services.export.AbstractExporterTest;
+
+public class ProgramApplicationsApiControllerTest extends AbstractExporterTest {
+
+  private ApiKey apiKey;
+  private static final String keyId = "key-id";
+  private static final String keySecret = "key-secret";
+  private static final String rawCredentials = keyId + ":" + keySecret;
+  private static final String serializedApiKey =
+      Base64.getEncoder().encodeToString(rawCredentials.getBytes(StandardCharsets.UTF_8));
+
+  @Before
+  public void setUp() {
+    apiKey = resourceCreator.createActiveApiKey("test-key", keyId, keySecret);
+  }
+
+  @Test
+  public void list_success_dateRange() {
+    createFakeProgramWithEnumerator();
+    apiKey
+        .getGrants()
+        .grantProgramPermission(fakeProgramWithEnumerator.getSlug(), ApiKeyGrants.Permission.READ);
+    apiKey.save();
+
+    String requestUrl =
+        controllers.api.routes.ProgramApplicationsApiController.list(
+                fakeProgramWithEnumerator.getSlug(),
+                /* fromDate= */ Optional.of("2022-01-15"),
+                /* toDate= */ Optional.of("2022-02-15"),
+                /* nextPageToken= */ Optional.empty(),
+                /* pageSize= */ Optional.empty())
+            .url();
+
+    Result result = doRequest(requestUrl);
+    assertThat(result.status()).isEqualTo(HttpStatus.SC_OK);
+
+    DocumentContext resultJson = JsonPathProvider.getJsonPath().parse(contentAsString(result));
+    assertThat(resultJson.read("payload[0].application_id", Long.class))
+        .isEqualTo(applicationTwo.id);
+    assertThat(resultJson.read("payload.length()", Integer.class)).isEqualTo(1);
+  }
+
+  @Test
+  public void list_success_multiplePages() {
+    createFakeProgramWithEnumerator();
+    apiKey
+        .getGrants()
+        .grantProgramPermission(fakeProgramWithEnumerator.getSlug(), ApiKeyGrants.Permission.READ);
+    apiKey.save();
+
+    String firstRequestUrl =
+        controllers.api.routes.ProgramApplicationsApiController.list(
+                fakeProgramWithEnumerator.getSlug(),
+                /* fromDate= */ Optional.empty(),
+                /* toDate= */ Optional.empty(),
+                /* nextPageToken= */ Optional.empty(),
+                /* pageSize= */ Optional.of(2))
+            .url();
+
+    Result result = doRequest(firstRequestUrl);
+    assertThat(result.status()).isEqualTo(HttpStatus.SC_OK);
+
+    DocumentContext resultJson = JsonPathProvider.getJsonPath().parse(contentAsString(result));
+    assertThat(resultJson.read("payload[0].application_id", Long.class))
+        .isEqualTo(applicationThree.id);
```

Comment:
Related to the above, this magic is fragile.  Can we make the ordering constraint a clear part of the setup, or if there is an underlying invariant in the setup, comment that for readers of the test.

Reply:
done

Reply:
Added javadoc for `createFakeProgramWithEnumerator`

Reply:
For here, comment at the start that there are three applications in reverse creation order to account for relevant test "setup" information.

Reply:
done